### ISSUE TITLE
archive-push taking long affecting performance of postgresql?

### ISSUE THREAD

**Issue by martindomeij** [pgbackrest/pgbackrest#1700]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.37

2. PostgreSQL version: 11.9

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each: Ubuntu 18.04

4. Did you install pgBackRest from source or from a package? Package

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
pg_primary:
[global]
repo1-host=repo1
repo2-host=repo2
archive-async=y

7. Describe the issue:
Before I started to implement pgbackrest, I would have archive_command set to a shell script that would basically use cp to copy the wal file to a local location. I would then use a clustered file system to get the files to the two backup servers, 'in the background', so to speak. archive_command would finish in a matter of a couple of tenths of a second at most.

Using pgbackrest to push the wal file to two backup servers (over ssh) will take anything from 1.2 seconds to 3.5 seconds, depending on how I set compression, if I reuse the ssh connection, etc. In other words, it now takes at least a second longer before archive_command finishes.

I am trying to grasp if or how this difference in time would or could affect the performance of postgresql, but I can't really find anything in the documentation.

Would it make any sense at all to have archive_command cp to local disk and then have pgbackrest read the file from there and push it to repos, just to have it out of the hands of postgresql? Or would that just be stupid?

Postgresql writes a wal file of 16 MB about every ten seconds during slow hours, and perhaps one every 1-2 seconds during peak hours. Pgbackrest does seem to be able to keep up though, for now at least.

**Comment by pgstef** [comment]
Hi,

> I am trying to grasp if or how this difference in time would or could affect the performance of postgresql, but I can't really find anything in the documentation.

AFAICT, a "slow" archive_command don't affect PG performance. Server-side resource consumption (for compression,...) can perhaps impact it but that would have to be measured. If compression is slowing too much (or taking too much resource), you can change the compress-type or even the compress-level.

What could happen if the archive_command is too slow compared to PG generating WAL segments is that PG won't recycle/remove those WAL segments not archived yet and you could have a big disk space used by those. 
Usually when archive_command is too slow (for any reason, like ssh connection latency), using `process-max` + `archive-async` really helps speeding things up (and even bring fault-tolerance when using multiple-repositories).

> Would it make any sense at all to have archive_command cp to local disk and then have pgbackrest read the file from there and push it to repos, just to have it out of the hands of postgresql? Or would that just be stupid?

Wouldn't make any sense imo. Just use async archiving then, that's exactly done for that.
If ssh is slowing you down too much, you could also try the latest feature -> TLS server. It might help speeding things up.

Kind Regards

**Comment by martindomeij** [comment]
Thank you for your reply.

I have already set archive-async=y, so I can't improve anything there.

I have tried setting process-max to 2, 5, and 10 without seeing any difference in time. If anything, a setting of 10 seems to add a few tenths of a second to the push.

I have also tried different compress-level/compress-type settings and found a sort of a sweet spot with 3/zst.

I read about the TLS server and I might try it. I have however assumed that most of the time consumed by ssh is setting up the connection, and by reusing them, I should be better off. But perhaps the TLS server is faster still?

**Comment by pgstef** [comment]
> But perhaps the TLS server is faster still?

I've not compared it myself but indeed heard that a lot of performance tests have shown TLS to be significantly faster than SSH.
If the SSH connection is your choke point, that would be a good idea to try I think :thinking:

**Comment by dwsteele** [comment]
pgBackRest will be faster the more WAL you are generating. At the rates you are talking about, there will generally only be one WAL file to push at any time. That's why increasing process-max does not help. I would set process-max at 2-4 and monitor archiving to be sure it does not fall behind in the future if the database gets busier.

**Comment by martindomeij** [comment]
> I've not compared it myself but indeed heard that a lot of performance tests have shown TLS to be significantly faster than SSH. If the SSH connection is your choke point, that would be a good idea to try I think

This sounds reasonable. Unfortunately, I can't seem to get tls to work. I'm pulling my hair...

2022-03-22 16:58:54.065 P00  ERROR: [103]: unable to find a valid repository:
                                    repo1: [AccessError] raised from remote-0 tls protocol on 'repo2': access denied
                                    repo2: [AccessError] raised from remote-0 tls protocol on 'repo1': access denied
                                    --------------------------------------------------------------------
                                    If SUBMITTING AN ISSUE please provide the following information:

                                    version: 2.37
                                    command: archive-push:async
                                    options: [/mnt/data/postgresql/11/demo/pg_wal] --archive-async --archive-push-queue-max=500G --compress-level=3 --exec-id=21780-17444964 --log-level-console=off --log-level-file=debug --log-level-stderr=off --pg1-path=/mnt/data/postgresql/11/demo --repo1-host=repo1 --repo2-host=repo2 --repo1-host-ca-file=/etc/pgbackrest/cert/ca.crt --repo2-host-ca-file=/etc/pgbackrest/cert/ca.crt --repo1-host-cert-file=/etc/pgbackrest/cert/repo1.crt --repo2-host-cert-file=/etc/pgbackrest/cert/repo2.crt --repo1-host-key-file=/etc/pgbackrest/cert/repo1.key --repo2-host-key-file=/etc/pgbackrest/cert/repo2.key --repo1-host-type=tls --repo2-host-type=tls --spool-path=/mnt/data/pgbackrest/spool --stanza=demo

repo2 has the following configuration:
[global]
repo2-path=/backup/local
tls-server-address=*
tls-server-auth=pg1=demo
tls-server-ca-file=/etc/pgbackrest/cert/ca.crt
tls-server-cert-file=/etc/pgbackrest/cert/repo2.crt
tls-server-key-file=/etc/pgbackrest/cert/repo2.key
log-level-file=debug

[demo]
pg1-host=pg1
pg1-port=5433
pg1-path=/mnt/data/postgresql/11/demo

pg1-host-type=tls
pg1-host-ca-file=/etc/pgbackrest/cert/ca.crt
pg1-host-cert-file=/etc/pgbackrest/cert/pg1.crt
pg1-host-key-file=/etc/pgbackrest/cert/pg1.key

**Comment by martindomeij** [comment]
> > But perhaps the TLS server is faster still?
> 
> I've not compared it myself but indeed heard that a lot of performance tests have shown TLS to be significantly faster than SSH. If the SSH connection is your choke point, that would be a good idea to try I think thinking

Indeed. Now down to 500-700 ms with tls. Thank you.

**Comment by martindomeij** [comment]
> pgBackRest will be faster the more WAL you are generating. At the rates you are talking about, there will generally only be one WAL file to push at any time. That's why increasing process-max does not help. I would set process-max at 2-4 and monitor archiving to be sure it does not fall behind in the future if the database gets busier.

Yes, that sounds reasonable. I will do this. And I did notice that it is a lot faster when there are many files. I just now pushed 16000 wal files to two repo servers in about 22 minutes using tls.
